### PR TITLE
Block Supports: Add text columns (column count) to typography block supports

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Related: 
- https://github.com/WordPress/gutenberg/pull/46566
- https://github.com/WordPress/gutenberg/issues/25459 
- https://github.com/WordPress/gutenberg/issues/27118

## What?

Adds a new block support feature to the typography toolset to allow control over CSS column count for blocks displaying multiline text.

## Why?

To open up further design possibilities.

### Patterns enabled by this

| Post Excerpt | Paragraph |
|---|---|
|  |  |

## How?
- Adds new "text columns" block support that maps to the `column-count` CSS property.
- The new block support and design tool is disabled by default (happy to change this if desired)
- Updates global styles typography panel to include new control if enabled.

## Testing Instructions

1. Run `npm run test:unit:php -- --filter WP_Theme_JSON_Gutenberg_Test`
2. Run `npm run test:unit packages/block-editor/src/hooks/test/style.js`
3. Enable the text columns UI by setting your theme.json's `settings.typography.textColumns` to `true`
4. Opt into text columns support for a block such as Post Excerpt by setting its `supports.typography.textColumns` property to `true`.
5. Edit a post, add your block with text columns support, add text to it, then select it.
6. In the typography panel in the sidebar, toggle on the text columns control via the panel's ellipsis menu
7. Adjust the control and ensure it works.
8. Save the post and confirm the columns display appropriately on the frontend.
9. Test text columns in global styles for blocks works.
10. Make sure the text columns control isn't an option under the Global Stlyes' root typography panel.

## Screenshots 

https://user-images.githubusercontent.com/60436221/207778088-553f8e08-6b0c-48f6-b432-e1ba6334f854.mp4

### CODE REVIEW

**Comment by ramonjd** [comment]
This is looking great, and, in my view, satisfies the feature requests in https://github.com/WordPress/gutenberg/issues/25459 and https://github.com/WordPress/gutenberg/issues/27118, though I'll leave it to @scruffian and @iamtakashi to judge.

I mostly tested with `core/paragraph` but also the post tags block as well.

1. Run the above tests and ensure they pass ✅ 
2. Enable column count support by opting in via your theme.json ✅ 
3. Test a static block such as the paragraph or post excerpt blocks. ✅ 
4. Test a dynamic block such as the post-terms block. ✅ 
5. Test the columns support via Global Styles ✅ 
6. Test theme.json styling of columns ✅ 


I think it's good to go.

The only question I had was whether we should clear the value in the UI when we reset the column count to make it clearer to the user that no column count is specified (and therefore may be overridden by global styles).

Dragging the column count down to `1` on a paragraph for example, renders the following:

```html
<p style="column-count:1">...</p>
```

The consequence was that global styles didn't apply to this particular paragraph:

Screen Shot 2021-07-22 at 2 08 49 pm

I would expect this to be the case. I think it makes sense to be able to specify that a block should always have `n` columns, global styles notwithstanding 👍 

Still, since hitting "Reset" changes the UI value to `1` it's not immediately clear that _resetting_ to `1` allows global style overrides, while manually adjusting the value to`1` does not.

I could be overthinking it.

**Comment by aaronrobertshaw** [comment]
Thanks for the review @ramonjd 🙇 

> The only question I had was whether we should clear the value in the UI when we reset the column count to make it clearer to the user that no column count is specified (and therefore may be overridden by global styles).

That's a good question. The current behaviour comes from setting the initial position on the slider control. It might be better to omit that so the input can reflect the "unset" state indicating that it's currently inheriting from the theme or global styles. 

Originally, I was thinking along the lines of the columns control for the gallery block that was currently a slider which always displayed a value. There could also be a case to change this to a stepper field.

In the short term, I'll try and improve its behaviour without switching controls. Then wait on further feedback regarding the UX.

> I could be overthinking it.

I don't think so. It's definitely a rough edge that could do with smoothing out 🙂

**Comment by aaronrobertshaw** [comment]
The column count's `RangeControl` has had the `initialPosition` prop removed. This does make it clearer when the control has been reset and would be inheriting from the theme or global styles. The downside is the position of the slider feels a bit off. 

When the new [progressive disclosure panel](https://github.com/WordPress/gutenberg/pull/32392) lands and the typography block supports panel is updated to use it, we'll likely want to reduce the control's footprint. It also won't need the reset button. I'm starting to lean more towards changing this to a simple `NumberControl`.

| RangeControl | NumberControl |
|---|---|
| ColumnControlResetBehaviour | ColumnControlResetBehaviour2 |

**Comment by ramonjd** [comment]
> The column count's RangeControl has had the initialPosition prop removed. This does make it clearer when the control has been reset and would be inheriting from the theme or global styles. 

Thank you!

> When the new progressive disclosure panel lands and the typography block supports panel is updated to use it, we'll likely want to reduce the control's footprint. It also won't need the reset button.

Makes sense to me, particularly after having just taken that PR for a spin.

Speaking _for_ the reset button, it is more explicit in that you know what's going to happen when you click on it.

As we continue to test the panel and controls like this one, maybe we can work out a neat way to communicate that values are reset/won't persist when toggling controls in the progressive disclosure panel.

>  I'm starting to lean more towards changing this to a simple NumberControl.

Nice idea 👍 More so given the range is 1-5. I'd understand the need for range controls if we were dealing with 0-100 with step values.

**Comment by mtias** [comment]
One challenge here is to not confuse users when it comes to Column blocks and this text columns setting. We'd need to be careful with how the UI controls are presented, and probably add further contextual text around what this does (cosmetically and not structurally, meaning a user won't be able to add other blocks to a column). For example, I'm not entirely sure this belongs in the typography panel.

cc @kellychoffman @javierarce @jasmussen

**Comment by mtias** [comment]
For example, maybe we can consider this more of a block style rather than a full on setting, and only support two columns.

**Comment by aaronrobertshaw** [comment]
Thanks for the feedback @mtias. 👍 

I'm happy to adjust this however needed to avoid any confusion with the existing Columns block.

While creating this PR I had a couple of things in mind:

1. That [multiple blocks with multi-line content might benefit](https://github.com/WordPress/gutenberg/issues/25459#issuecomment-699091152) from control over CSS columns and block support appeared the best method in providing that.
2. Providing control over this in the sidebar would indicate it was cosmetic as well as a point of difference between this and the columns block.
3. There could be the option in future that block styles are a set of style attributes. 

It appears I misunderstood what was initially being [proposed](https://github.com/WordPress/gutenberg/pull/33157#issuecomment-874658070) regarding block styles. I was under the impression the style attributes might also need to be supported via theme.json and/or block supports.

I don't have any strong opinions on where this should belong. As the discussion around this evolves I'll continue to make any requested updates.

**Comment by mtias** [comment]
Yes, I think the result is valuable to have, we just need to explore a bit the UI controls and what makes the most sense.

**Comment by javierarce** [comment]
>  I'm not entirely sure this belongs in the typography panel.

A "layout" panel would have more sense, in my opinion. Is there a plan to add more column styles like column-gap, column-rule, etc.?

> One challenge here is to not confuse users when it comes to Column blocks and this text columns setting.

I think that the fact that this setting offers fewer functionalities than the block could help users understand that this is the quick way to add columns, while the block is the tool to use if they need more control.

**Comment by aaronrobertshaw** [comment]
> A "layout" panel would have more sense, in my opinion. 

I'm not certain here. A layout panel is already injected for certain blocks to control content width and full width values etc. If we are trying to keep the delineation clear between columns blocks and other more structural changes, it feels more typographic to me.

> Is there a plan to add more column styles like column-gap, column-rule, etc.?

Some additional control over styles is on the horizon. https://github.com/WordPress/gutenberg/pull/32571 adds the ability to control gap spacing and this would be included under the new "Dimensions" panel.

**Comment by jasmussen** [comment]
As the group block is getting more flex-features, and is even growing closer to the columns block, it seems like there's an opportunity to share some DNA in the interface, perhaps even in the underlying code. After all, what is the major difference between a 1 column Columns block, and a Group block? (I know there are some exotic alignment based ones).

The thing is, I share Matías concern about the confusion. Moreover, I have bouts of excitement about the built-in _column-count_ features, then I implement them and remember the downsides:

column-count

Specifically, box-shadows and borders, anything that's on the outside of an element, those are broken across columns. There's little to no control of it, and you can even end up in situations where the preceeding element's bottom margin pushes down the next column so it doesn't start at the top. You can tinker with CSS to mitigate things with `inline-block` and other tweaks, but in my experience I could never get it to a satisfactory place. You can tinker in [this codepen if you like](https://codepen.io/joen/pen/yLbqpXd?editors=1100).

In other words, I find `column-count` to be over-promising and under-delivering. If we are to implement it, we should manage the prominence so it doesn't end up just frustrating users with its built-in shortcomings. That could mean, as Matías suggests, making it a somewhat locked-down block style we know to work most of the time, or as Aaron suggests, a control you can add via the Tools menu, but which isn't ever shown by default.

**Comment by aaronrobertshaw** [comment]
Thanks for highlighting the issues with CSS columns @jasmussen . 

> That could mean, as Matías suggests, making it a somewhat locked-down block style we know to work most of the time, or as Aaron suggests, a control you can add via the Tools menu, but which isn't ever shown by default.

Having `column-count` supported through block supports or theme.json etc doesn't mean we are obligated to enable it for all blocks. It could be only turned on for the post excerpt and paragraph. The block support approach would mean that we have consistency between those blocks instead of two ad hoc solutions needing to be maintained separately.

It comes down to whether there are enough blocks that would benefit from it, that also don't need borders or shadows, to warrant it being implemented in a shared way such as this PR does. 

We could still switch to offering it as an ad hoc solution for the post excerpt block etc. In that case, it would be good to offer a `SlotFill` solution in our different inspector control panels. This would allow the ad hoc columns count control to be injected within the typography panel. Not to mention potentially facilitate a more general "Appearance" panel spanning multiple block supports.

Also, for what it's worth, I still think typesetting is different than layout and so the CSS columns fit better in the typography panel than the layout one. Happy to be overruled there, it's a minor issue in the scheme of things.

**Comment by jasmussen** [comment]
> Having column-count supported through block supports or theme.json etc doesn't mean we are obligated to enable it for all blocks. It could be only turned on for the post excerpt and paragraph. The block support approach would mean that we have consistency between those blocks instead of two ad hoc solutions needing to be maintained separately.

The tools menu could offer the choice, that's the benefit of it! Though per the feedback I think we should at the very least start with it not shown by default.

**Comment by ZebulanStanphill** [comment]
@jasmussen I figured I should mention that the issue with borders/shadows being split across columns can be addressed via the [`break-inside`](https://developer.mozilla.org/en-US/docs/Web/CSS/break-inside) CSS property.

https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Multiple-column_Layout#fragmented_boxes

**Comment by mtias** [comment]
Let's pick this one up and drive it to the end this time :)

File: lib/class-wp-theme-json-gutenberg.php
Comment (mtias) [inline lib/class-wp-theme-json-gutenberg.php]: Can we change the attribute to be `textColumns` instead?

**Review by mtias — DISMISSED** [review]
Left a request to change the attribute and function names to `textColumns` instead of `columnCount`.

We could explore a different UI with a segmented control and icons for single column, two-columns, and three columns. I don't think we need to support more than that for text-columns.

File: lib/class-wp-theme-json-gutenberg.php
Comment (ZebulanStanphill) [reply lib/class-wp-theme-json-gutenberg.php]: Well, technically, it's not _just_ columns for text: something using this property could have images and other media in its content.

In fact, are we even sure this belongs in the category of typography tools? It technically has use cases outside of bare text, especially once you start using `break-inside` to preserve the integrity of child containers. A gallery could hypothetically use it to layout its images, for example, as could a series of listing of social media posts. I'd argue it actually belongs with the layout tools.

I think "content columns" is a more strictly accurate term than "text columns".

Comment (mtias) [reply lib/class-wp-theme-json-gutenberg.php]: Well, the original issues are about unlocking a specific effect on textual content. I'm not sure we should design it as a general tool and expose it as one were it'd start to compete with other layout tools. "Content columns" is just very close to regular columns, which are also about showing any content in columns. Here we need to represent the flow and continuity aspect, which seems better captured by "text columns" to me. It might be better placed in "layout" regardless, but with the right context.

There's also a question on which blocks this should be exposed. For example, wouldn't make any sense to have this on the headings block.

Comment (ZebulanStanphill) [reply lib/class-wp-theme-json-gutenberg.php]: The interesting thing about CSS `columns` versus typical divs-in-divs "columns" is that:

- Divs-in-divs columns have content explicitly assigned to a column, and the height of each column can vary wildly.
- CSS `columns` have content dynamically placed in the columns, often attempting to balance the height of each column (though a minimum content height for a column can be set).

These have big ramifications for when you would want to use one over the other, but I have no clue how to convey these subtle but important differences to users.

---

Here's the only blocks I can think of where this control would make sense:

- Buttons
- Gallery
- Group
- List
- Navigation
- Paragraph
- Quote
- Verse

In other words: most of the container blocks, and _maybe_ a few standard text blocks for the sake of allowing users to avoid extra divs in a few cases. In most cases, though, you'd actually want to avoid using CSS `columns` on a single Paragraph, since that's less future-proof for adding additional text. So maybe it should only be the container blocks that have the control, to encourage best practices.

**Comment by aaronrobertshaw** [comment]
Thank you for the renewed conversation and interest in this @mtias and @ZebulanStanphill.

I'll dust this off, bring it up-to-date and rename functions/attributes etc, later this week as I'm AFK tomorrow.

**Comment by aaronrobertshaw** [comment]
Quick Update:
- I've created a new PR to [add support for the `column-count` CSS property as "text columns" to the style engine](https://github.com/WordPress/gutenberg/pull/46566)
- This PR is now based upon [#46566](https://github.com/WordPress/gutenberg/pull/46566) and has been updated to the follow the latest approaches to block supports and global styles

Unfortunately, I'll be AFK until the new year. I'll try to keep an eye on this PR however anyone is welcome to pick it up if they'd like.

**Comment by aaronrobertshaw** [comment]
The addition of `column-count` to the style engine has been merged (https://github.com/WordPress/gutenberg/pull/46566).

It was [raised there](https://github.com/WordPress/gutenberg/pull/46566#pullrequestreview-1218657155) that some user agents may add some default `-webkit-margin-before` etc that results in some misalignment of the columns. The resulting effect differs a little based on whether the text columns are applied on the block wrapper or a specific inner `p` etc.

I'll explore this when I return from AFK.

If it turns out that applying the style to the correct element is sufficient, then this issue might not require any further changes and it would be handled when a given block adopts this support. An alternative option would be to update the style engine to return a `has-text-columns` class when this support is applied. That could then be used to apply a generic style reset or leverage as needed by blocks.

**Comment by aaronrobertshaw** [comment]
> I'll explore this when I return from AFK.

After digging into the user agent margin issue a little, I'm not convinced adding a `has-text-columns` class is a silver bullet. We'd still need to be either applying it to the correct location if not the wrapper (same as skipping serialization and applying the style), or adding potentially brittle margin overrides for inner elements.

I think the best option is for blocks adopting the block support to apply it as needed to achieve their desired result.  We could still add the `has-text-columns` class if people feel it would be useful.

For theme.json, the extended block.json selectors API proposed in https://github.com/WordPress/gutenberg/pull/46496 will provided fine-grained control over the selector the `column-count` style is generated for. 

With both individual blocks and theme.json/global styles covered, I think we can progress to reviewing this with a plan to land it soon.

**Comment by jasmussen** [comment]
If the CSS class addition can't be the sole arbiter of the columns, and we still need inline styles, I'd personally lean towards omitting the class then, just to keep things as lean as they can be.

**Comment by andrewserong** [comment]
Nice work sticking with this one, and it's testing pretty well for me so far! I was wondering if it's worth adding this control to the appearance tools opt-in so that themes don't need to explicitly declare support for it? Or is it experimental enough that it's good for it to be an explicit opt-in?

One thing I was wondering about — on narrower viewport widths, the columns can get a bit narrower, e.g. here's a three column layout (looks great on desktop, but is a little cluttered on mobile):

| Mobile | Desktop |
| --- | --- |
| image | image |

I haven't played with these CSS properties much before, but [from the sounds of it](https://css-tricks.com/almanac/properties/c/column-width/), we can pair `column-count` with `column-width` so we could define a width for the column to make the columns play a bit nicer on mobile. One idea might be to inject a `column-width` value into the markup at render time (like we do with fluid typography rules) so that we could use a magic value of some kind that plays nicely with the viewport size, without having to store that in post content? Just an idea, and not necessarily a blocker! I'd be curious to see how we can get it to play nicely across different screen sizes, though 🤔 

Overall, though, I'm pretty excited to see text columns support added — sounds like it's going to unlock some nice magazine-style layouts!

**Comment by aaronrobertshaw** [comment]
Thanks for the review @andrewserong 🙇 

> I was wondering if it's worth adding this control to the appearance tools opt-in so that themes don't need to explicitly declare support for it?

My understanding was that we wanted to be pretty selective with which blocks may adopt this support. That made me err more on the side of this support not being included within the appearance tools auto-opt-ins. If it does prove desirable, it would be an easy follow-up to add it there. Any themes already opting in manually, wouldn't be negatively impacted.

> One thing I was wondering about — on narrower viewport widths, the columns can get a bit narrower, e.g. here's a three column layout (looks great on desktop, but is a little cluttered on mobile):
> ...
> we can pair column-count with column-width so we could define a width for the column to make the columns play a bit nicer on mobile.

Sounds like a great follow-up. I think it would be best to land this new support, then that could even be worked on while we explore which blocks would benefit most from text columns.

File: packages/edit-site/src/components/global-styles/typography-panel.js
Comment (andrewserong) [inline packages/edit-site/src/components/global-styles/typography-panel.js]: This line is throwing an error — it looks like after the rebase, this needs to be `useGlobalSetting` now?

**Review by andrewserong — COMMENTED** [review]
> My understanding was that we wanted to be pretty selective with which blocks may adopt this support. That made me err more on the side of this support not being included within the appearance tools auto-opt-ins.

That sounds like good logic to me — it means we could safely land the block support and be deliberate about when to automatically opt the appearanceTools into the support, possibly once there's column width happening, too. And like you mention, it means only themes that know and are happy with the current behaviour of text columns will get it 👍

> Sounds like a great follow-up. I think it would be best to land this new support, then that could even be worked on while we explore which blocks would benefit most from text columns.

Agreed, especially if we wind up pursuing a dynamic approach to injecting column width values, there's nothing there that should block this PR IMO.

File: packages/edit-site/src/components/global-styles/typography-panel.js
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/global-styles/typography-panel.js]: Thanks for taking a look. I rushed the rebase as I was about to run late for a meeting. I'll fix that up now.

Comment (aaronrobertshaw) [reply packages/edit-site/src/components/global-styles/typography-panel.js]: ✅  Done in https://github.com/WordPress/gutenberg/pull/33587/commits/6a7de1c4f851e13766adec164f8a1cd1a3a0b797

Comment (aaronrobertshaw) [reply packages/edit-site/src/components/global-styles/typography-panel.js]: The variations path change was also missed after the rebase. I've updated that as well.

**Comment by aaronrobertshaw** [comment]
Given the evolving nature of this block support (i.e. `column-width` potentially being injected at runtime), I was wondering if we'd be better served landing this after the 6.2 feature freeze and iterating it from there within the Gutenberg plugin. What do you think @andrewserong?

> We could explore a different UI with a segmented control and icons for single column, two-columns, and three columns. I don't think we need to support more than that for text-columns.

Delaying to post 6.2 would also provide some time to explore the segmented control before releasing this into the wild.

**Comment by andrewserong** [comment]
> I was wondering if we'd be better served landing this after the 6.2 feature freeze and iterating it from there within the Gutenberg plugin. What do you think @andrewserong?

If you're happy with that, I'd be in favour of landing this after the 6.2 feature freeze, as it'd give us a bit more time and flexibility to experiment with dynamic output for width, etc, as you mention. So long as there's no particular urgency from anyone to get this out sooner, sounds good to me!

**Comment by aaronrobertshaw** [comment]
While it would be great to land it, I don't think it warrants distracting from higher priorities for 6.2. 

Several block support related APIs need stabilizing, and in some ways, it would make sense to achieve that before landing this more niche block support.

Happy to revisit this though, if others disagree.

**Comment by aaronrobertshaw** [comment]
I've rebased this PR to bring it up-to-date. 

In addition, I've updated the text columns support to use a [segmented control](https://github.com/WordPress/gutenberg/pull/33587#pullrequestreview-1103110952), leveraging the `ToggleGroupControl` component.

While this works well, now that the component offers the ability to deselect an option, it is different from the text-decoration and letter case controls. It would be great to get some design feedback on which option is preferred here.

Do we convert the text-decoration and letter case controls to use the `ToggleGroupControl`, or do we switch the text columns control to use a custom button group? 

<details>
<summary><strong>Here's a quick demo of the two types of controls side-by-side</strong></summary>

https://user-images.githubusercontent.com/60436221/216933232-6108e3a5-05f1-446c-ae0a-71f162841a1e.mp4

</details>

For the control's icons, I've temporarily used the column icons from the Column block's placeholder. I can replace these with more appropriate icons or add them to the icon library.

@jasmussen do you have any preferences here?

**Comment by jasmussen** [comment]
Nice work. I'm curious about the best interface for text-columns, and would appreciate insights by @WordPress/gutenberg-design. To me it definitely seems clear that:

* The control should be non-default, added from the tools menu
* Take up a half column

Whether it should be a segmented control, I'm less sure about. I can certainly get you some icons that would work, and I tend to agree 3 is probably a good max for text columns. An alternative would be a stepper:

Screenshot 2023-02-06 at 10 48 53

That is to say, this plus/minus interface hasn't yet been implemented in the WordPress component set, I believe the stepper currently has just a tiny up/down arrow set in the input. But it might still be a basic interface for it?

**Comment by jameskoster** [comment]
A question that often comes up with these controls is whether or not they need to accommodate an "Unset" / "Inherit" / "Default" value. This covers cases where a stylesheet adds a text columns style, and you want to remove it in the Editor. Yes, you could kind of do this by applying a specific value using this control, but then drawback to that approach is that it doesn't 'respect' default values in global styles. 

We need to address this holistically in https://github.com/WordPress/gutenberg/issues/43082, but for now, if we feel this control _does_ need such an affordance then that might help us determine which UI is best.

**Comment by aaronrobertshaw** [comment]
Thanks for the input @jasmussen and @jameskoster 🙇 

> That is to say, this plus/minus interface hasn't yet been implemented in the WordPress component set

We do actually have the custom spin buttons for the `NumberControl` within `@wordpress/components` 🎉 

I've taken the liberty to update this PR to use the `NumberControl` as a stepper. The following videos illustrate how the stepper looks vs the segmented control.

| Stepper (checkout latest commit) | Segmented Control (checkout https://github.com/WordPress/gutenberg/pull/33587/commits/72703988d669f02af227a3db7f9cb053f4a25feb) |
|---|---|
| <video src="https://user-images.githubusercontent.com/60436221/217177535-991e3277-4201-4c50-b47b-c87971e2161e.mp4" /> | <video src="https://user-images.githubusercontent.com/60436221/217177605-37aea69f-38dc-4c41-b284-68a2c71ce939.mp4" /> |

Personally, I like the stepper. What do you all think?

**Comment by jasmussen** [comment]
Oh nice, that plus to add columns, and minus to remove, that feels really ergonomic. Let's try it. Does the control support min/max-values? Not blocking, but we could limit the range from 0-6, just to pick a number.

**Comment by paaljoachim** [comment]
Ooooh this is nice! I really like the stepper plus/minus to add or remove a column! It really becomes nicely integrated into the Typography controls.

**Comment by aaronrobertshaw** [comment]
Thanks for the positive feedback @jasmussen nad @paaljoachim 👍 

Sounds like the stepper is a winner.

>  Does the control support min/max-values? Not blocking, but we could limit the range from 0-6, just to pick a number.

Yes, the `NumberControl` does support min/max values. I originally set the max to 5 but dropped it to 3 when trialling the segmented control. 

I've just updated the min/max range to be 1-6 in line with your suggestion. Zero columns doesn't seem to fit the use case here so went with the minimum of one.

This should be in a decent shape to land now I think. We can explore adding the `column-width` to the mix to better support columns on mobile displays in a follow-up.

**Comment by jasmussen** [comment]
> I've just updated the min/max range to be 1-6 in line with your suggestion. Zero columns doesn't seem to fit the use case here so went with the minimum of one.
> This should be in a decent shape to land now I think. We can explore adding the column-width to the mix to better support columns on mobile displays in a follow-up.

Sounds good, yeah we can always adjust these values upon learning. I suggested 6 mainly to support use cases that don't immediately come to mind, and to allow people to go just above what seems feasible and then pull back as they preview thre result.

**Comment by paaljoachim** [comment]
I am wondering...
Is 0 not to use columns?
1 or higher to use columns?
I am asking as someone might add the control and try it out and decide not to use it anyway. Would the columns be removed as in 0?

**Comment by jasmussen** [comment]
My assumption is that 1 means 1 text column, which is the default for any element. So I would assume when adding this control, 1 is the preset, and lowest value.

**Comment by aaronrobertshaw** [comment]
> My assumption is that 1 means 1 text column, which is the default for any element.

That's my thinking.

> Is 0 not to use columns?

As text columns is an optional control that needs to be toggled on from the typography panel's menu. It would be reset via that same menu to opt out of setting any specific column number. The field can also be cleared without resetting via the menu.

**Comment by paaljoachim** [comment]
Thank you for clearing that up.
1 is the lowest number and default for any element.

**Comment by aaronrobertshaw** [comment]
I've reworked this PR in line with the recently extracted typography panel https://github.com/WordPress/gutenberg/pull/47356. Given the degree of change required, this PR will need a full review with particular attention on whether the UI control displays appropriately based on theme.json settings or individual block supports.

**Comment by glendaviesnz** [comment]
Testing well for me. I think that after [this PR](https://github.com/WordPress/gutenberg/pull/46750) the `theme.json` changes are just supposed to go into the `lib/theme.json` file rather than in the versioned dir, but I am not 100% sure on that.

**Comment by andrewserong** [comment]
Thanks for the re-review ping, I haven't had a chance to test this properly yet, but just on this comment:

> I think that after https://github.com/WordPress/gutenberg/pull/46750 the theme.json changes are just supposed to go into the lib/theme.json file rather than in the versioned dir, but I am not 100% sure on that.

I think that's the case, now that the resolver isn't in the versioned directory, it looks like the `lib/theme.json` file is the one to use 👍

**Comment by aaronrobertshaw** [comment]
Thanks @glendaviesnz and @andrewserong for catching the incorrect theme.json. I'd lost track of this during the multitude of rebases after significant changes to theme.json/global styles panels etc.

I've moved the theme.json tweak into `lib/theme.json` and re-tested this branch locally. Works for me.

File: packages/block-editor/src/hooks/text-columns.js
Comment (andrewserong) [inline packages/block-editor/src/hooks/text-columns.js]: Is this file still needed?

File: packages/block-editor/src/components/global-styles/typography-panel.js
Comment (andrewserong) [inline packages/block-editor/src/components/global-styles/typography-panel.js:426]: Since this uses the `NumberControl` component, do we still need the separate `TextColumnsEdit` component?

**Review by andrewserong — APPROVED** [review]
Thanks for all the follow-up here @aaronrobertshaw, this is testing great for me!

✅ Works correctly at the individual block level for both static and dynamic blocks (e.g. post excerpt, paragraph blocks)
✅ Works in global styles / theme.json, and can be overridden at the individual block level
✅ Control doesn't show up where it shouldn't (e.g. root level text elements in global styles)
✅ Switching `textColumns` to `false` in `theme.json` still allows the text columns styles to be displayed in the block editor and site frontend for blocks that already have those values set (e.g. in patterns or existing post content when a user switches themes)

From a re-read, the code's looking solid to me. The only comments I added is that after the rebase now that the Typography panel is consolidated between the block editor and site editors, is that it looks like we might be able to remove the `text-columns.js` file? I couldn't see it being imported, since `typography-panel.js` now has the logic in there directly using `NumberControl`, but apologies if I've missed something there!

LGTM! ✨

File: packages/block-editor/src/hooks/text-columns.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/text-columns.js]: 🤦 Fixed.

File: packages/block-editor/src/components/global-styles/typography-panel.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/global-styles/typography-panel.js:426]: No. Thanks 👍

**Comment by aaronrobertshaw** [comment]
Appreciate the speedy reviews, thank you 🚀 

> apologies if I've missed something there!

Nope, the opposite, I missed it along with the `lib/theme.json` change.

It was kept around for reference when I was rebasing after the typography panel extraction. I then dropped the ball in terms of removing it once it wasn't needed.

With all the feedback addressed, I'll go ahead and merge this one now.

**Comment by andrewserong** [comment]
> It was kept around for reference when I was rebasing after the typography panel extraction. I then dropped the ball in terms of removing it once it wasn't needed.

No worries, happy to be another set of 👀, it's tricky to see these things sometimes when you've been with a PR for a while 😄. Nice work landing this feature!

**Comment by ramonjd** [comment]
I'll backport the PHP files for this one 👍

**Pull request merged**
